File: ReSharper.FSharp/src/Psi.FSharp.LanguageService/src/FSharpPsiBuilderTokenFactory.cs
using JetBrains.ReSharper.Psi.ExtensionsAPI.Tree;
using JetBrains.ReSharper.Psi.FSharp.Impl.Tree;
using JetBrains.ReSharper.Psi.FSharp.Parsing;
using JetBrains.ReSharper.Psi.Parsing;
using JetBrains.ReSharper.Psi.TreeBuilder;
using JetBrains.Text;

namespace JetBrains.ReSharper.Psi.FSharp.LanguageService
{
  public class FSharpPsiBuilderTokenFactory : IPsiBuilderTokenFactory
  {
    public LeafElementBase CreateToken(TokenNodeType tokenNodeType, IBuffer buffer, int startOffset, int endOffset)
    {
      return tokenNodeType == FSharpTokenType.IDENTIFIER || tokenNodeType == FSharpTokenType.OPERATOR
        ? new FSharpIdentifierToken(tokenNodeType, buffer, new TreeOffset(startOffset), new TreeOffset(endOffset))
        : tokenNodeType.Create(buffer, new TreeOffset(startOffset), new TreeOffset(endOffset));
    }
  }
}

File: ReSharper.FSharp/src/Psi.FSharp.Parsing/FSharpPsiBuilderTokenFactory.fs
namespace JetBrains.ReSharper.Psi.FSharp.Parsing

open JetBrains.ReSharper.Psi.TreeBuilder
open JetBrains.ReSharper.Psi
open JetBrains.ReSharper.Psi.FSharp.Impl.Tree
open JetBrains.ReSharper.Psi.ExtensionsAPI.Tree

type FSharpPsiBuilderTokenFactory() =
    interface IPsiBuilderTokenFactory with
        member this.CreateToken(tokenType, buffer, startOffset, endOffset) =
            let startOffset = TreeOffset(startOffset)
            let endOffset = TreeOffset(endOffset)
            
            if obj.ReferenceEquals(tokenType, FSharpTokenType.IDENTIFIER) ||
               obj.ReferenceEquals(tokenType, FSharpTokenType.OPERATOR)
            then FSharpIdentifierToken(tokenType, buffer, startOffset, endOffset) :> LeafElementBase
            else tokenType.Create(buffer, startOffset, endOffset)